src/vector/avx2.rs

```diff
@@ -164,24 +151,35 @@ impl u8x32 {
             // TL;DR avx2's PALIGNR instruction is actually just two 128-bit
             // PALIGNR instructions, which is not what we want, so we need to
             // do some extra shuffling.
-            let v = _mm256_permute2x128_si256(other.vector, self.vector, 0x21);
-            let v = _mm256_alignr_epi8(self.vector, v, 15);
-            u8x32 { vector: v }
+            let v = _mm256_permute2x128_si256(other.0, self.0, 0x21);
+            let v = _mm256_alignr_epi8(self.0, v, 15);
+            u8x32(v)
         }
     }
 
     #[inline]
     pub fn bit_shift_right_4(self) -> u8x32 {
         // Safe because we know AVX2 is enabled.
         unsafe {
-            u8x32 { vector: _mm256_srli_epi16(self.vector, 4) }
+            u8x32(_mm256_srli_epi16(self.0, 4))
         }
     }
+
+    #[inline]
+    pub fn bytes(self) -> [u8; 32] {
+        // Safe because __m256i and [u8; 32] are layout compatible
+        unsafe { mem::transmute(self) }
+    }
+
+    #[inline]
+    pub fn replace_bytes(&mut self, value: [u8; 32]) {
```

Comment:
Perhaps better to have `bytes(&self) -> &[u8; 32]` and `bytes_mut(&mut self) -> &mut [u8; 32]`? It would be more Rust-idiomatic than providing a getter+setter, and in some cases might give a faster way to change separate bytes without replacing the whole thing.

Reply:
> Perhaps better to have bytes(&self) -> &[u8; 32]

Where does that `&[u8; 32]` live ?

Reply:
The body of the function wouldn't change - you'd still use `transmute`, just between references and not values.

Reply:
Here's an example with both implementations and you can see how code for setting a single byte differs between two: https://rust.godbolt.org/z/XF7m95

Reply:
If the `__m256i` is in a SIMD register, it doesn't have a memory address, so creating a pointer to it requires spilling it to the stack to be able to give it an address. Once the modification on the stack is done, moving it back into a SIMD register requires copying the whole thing, not a single byte, particularly if dynamic indices are involved (if the indices are compile-time constant sometimes the compiler can do better, if there is a specific instruction for that).

One doesn't want this back-and-forth to happen every time one modifies a part of the vector. The proposed API forces its users to make the moves back to memory manually for the whole vector, and to copy the result manually back into the SIMD register. This revealed a couple of places where, e.g., instead of doing this back-and-forth on every iteration of a loop, one can just load the vector into memory once before the loop, operate on memory, and move the contents back into a SIMD register after the loop has completed.

Reply:
That's a good point. I'd argue that having several `.bytes_mut()[...] = ...` patterns in a row would also be pretty obvious and most people would store the reference when they could, but I see how your proposed API forces this a bit better.

However, if that's the goal, I wonder if it's worth to go one step further and take mutations as a callback like `.with_bytes(|bytes| ...)` where one can modify contents in any way they want inside, but then on the type system level you wouldn't be able to do anything else outside, and transformation back to a register would be guaranteed to happen as part of the call?

Reply:
> However, if that's the goal, I wonder if it's worth to go one step further and take mutations as a callback like .with_bytes(|bytes| ...) where one can modify contents in any way they want inside, but then on the type system level you wouldn't be able to do anything else outside, and transformation back to a register would be guaranteed to happen as part of the call?

Some code only wants to read the bytes, so forcing a write would be bad for that code - I've pinged you in one example.

I also don't see really an advantage in restricting access to a scope. If you wanted to mutate two vectors, you would need to nest scopes. For an API that's only intended for internal consumption, this API would feel even more like overengineering than the one I proposed - and I consider mine borderline overengineering.

Reply:
Okay, fair enough. I'm not 100% comfortable with `replace_bytes` vs `bytes_mut` for idiomaticity reasons, but I can see how this is the least of the evils :)

Reply:
> Here's an example with both implementations and you can see how code for setting a single byte differs between two: https://rust.godbolt.org/z/XF7m95

As mentioned, this should properly benchmarked before being merged. Those examples do not really use the intrinsics, so the code is being generated in isolation without any objective in mind, which does not represent what this library actually does. If you actually try to use the, you'll see that they generate the exact same code when all optimizations are turned on: https://rust.godbolt.org/z/Z8sdH9

But this is not about optimizing the implementation, this is about optimizing the amount of work that LLVM has to do to produce efficient code. If you look at the LLVM-IR produced in debug mode: https://rust.godbolt.org/z/sMfxHm the version using `replace_bytes` produces 134 lines of LLVM-IR, while the version using `bytes_mut` produces 758 lines of LLVM-IR. There are a couple of factors at play here, but producing 6x more LLVM-IR for this isn't really worth it.